File: sdks/csharp/Transports/Codec.cs
using MessagePack;

namespace Nexus.SDK.Transports;

/// <summary>
/// Wire codec for the Nexus RPC protocol.
///
/// The server-side types live in <c>nexus_protocol::rpc::types</c>.
/// rmp-serde encodes Rust enums using the externally-tagged
/// representation by default, which means:
///
/// <list type="bullet">
/// <item>Unit variants (<c>NexusValue::Null</c>) → the string <c>"Null"</c>.</item>
/// <item>Data-bearing variants (<c>NexusValue::Str("hi")</c>) → a single-key
///   map <c>{"Str": "hi"}</c>.</item>
/// <item><c>Result&lt;T, E&gt;</c> → <c>{"Ok": v}</c> / <c>{"Err": s}</c>.</item>
/// </list>
///
/// <see cref="ToWire"/> / <see cref="FromWire"/> translate between the
/// tagged .NET type and the on-wire shape.
/// <see cref="EncodeRequestFrame"/> / <see cref="DecodeResponseBody"/>
/// handle the <c>u32 LE length prefix + MessagePack body</c> frame format.
/// </summary>
public static class Codec
{
    /// <summary>Encode a <see cref="NexusValue"/> into its on-wire shape.</summary>
    public static object? ToWire(NexusValue v)
    {
        switch (v.Kind)
        {
            case NexusValueKind.Null:
                return "Null";
            case NexusValueKind.Bool:
                return new Dictionary<object, object?> { ["Bool"] = v.Value };
            case NexusValueKind.Int:
                return new Dictionary<object, object?> { ["Int"] = v.Value };
            case NexusValueKind.Float:
                return new Dictionary<object, object?> { ["Float"] = v.Value };
            case NexusValueKind.Bytes:
                return new Dictionary<object, object?> { ["Bytes"] = v.Value };
            case NexusValueKind.Str:
                return new Dictionary<object, object?> { ["Str"] = v.Value };
            case NexusValueKind.Array:
                var arr = (List<NexusValue>)v.Value!;
                var wireArr = new List<object?>(arr.Count);
                foreach (var e in arr) wireArr.Add(ToWire(e));
                return new Dictionary<object, object?> { ["Array"] = wireArr };
            case NexusValueKind.Map:
                var pairs = (List<(NexusValue Key, NexusValue Value)>)v.Value!;
                var wireMap = new List<object?>(pairs.Count);
                foreach (var (k, val) in pairs)
                    wireMap.Add(new List<object?> { ToWire(k), ToWire(val) });
                return new Dictionary<object, object?> { ["Map"] = wireMap };
            default:
                throw new InvalidOperationException($"unknown NexusValueKind '{v.Kind}'");
        }
    }

    /// <summary>Decode the wire-level shape back into a tagged <see cref="NexusValue"/>.</summary>
    public static NexusValue FromWire(object? raw)
    {
        switch (raw)
        {
            case null:
                return NexusValue.Null();
            case string s:
                return s == "Null" ? NexusValue.Null() : NexusValue.Str(s);
            case bool b:
                return NexusValue.Bool(b);
            case byte ub:
                return NexusValue.Int(ub);
            case sbyte sb:
                return NexusValue.Int(sb);
            case ushort us:
                return NexusValue.Int(us);
            case short ss:
                return NexusValue.Int(ss);
            case uint u32:
                return NexusValue.Int(u32);
            case int i32:
                return NexusValue.Int(i32);
            case ulong u64:
                return NexusValue.Int((long)u64);
            case long i64:
                return NexusValue.Int(i64);
            case float f32:
                return NexusValue.Float(f32);
            case double f64:
                return NexusValue.Float(f64);
            case byte[] bytes:
                return NexusValue.Bytes(bytes);
            case List<object?> list:
                {
                    var outArr = new List<NexusValue>(list.Count);
                    foreach (var e in list) outArr.Add(FromWire(e));
                    return NexusValue.Array(outArr);
                }
            case object[] arr:
                {
                    var outArr = new List<NexusValue>(arr.Length);
                    foreach (var e in arr) outArr.Add(FromWire(e));
                    return NexusValue.Array(outArr);
                }
            case IDictionary<object, object?> dict:
                return FromWireMap(dict);
            case IDictionary<string, object?> sDict:
                {
                    var convertedTag = new Dictionary<object, object?>();
                    foreach (var kv in sDict) convertedTag[kv.Key] = kv.Value;
                    return FromWireMap(convertedTag);
                }
        }
        throw new InvalidOperationException($"decode: unexpected NexusValue wire type {raw.GetType().Name}");
    }

    private static NexusValue FromWireMap(IDictionary<object, object?> dict)
    {
        if (dict.Count != 1)
            throw new InvalidOperationException(
                $"decode: expected single-key tagged NexusValue, got {dict.Count} keys");
        foreach (var kv in dict)
        {
            var tagStr = kv.Key as string ?? kv.Key.ToString();
            return FromTagged(tagStr!, kv.Value);
        }
        throw new InvalidOperationException("decode: empty tagged map");
    }

    private static NexusValue FromTagged(string tag, object? payload)
    {
        switch (tag)
        {
            case "Null": return NexusValue.Null();
            case "Bool": return NexusValue.Bool(Convert.ToBoolean(payload));
            case "Int": return NexusValue.Int(Convert.ToInt64(payload));
            case "Float": return NexusValue.Float(Convert.ToDouble(payload));
            case "Bytes":
                if (payload is byte[] b) return NexusValue.Bytes(b);
                if (payload is List<object?> lb)
                {
                    var bytes = new byte[lb.Count];
                    for (int i = 0; i < lb.Count; i++)
                        bytes[i] = Convert.ToByte(lb[i]);
                    return NexusValue.Bytes(bytes);
                }
                throw new InvalidOperationException("decode: Bytes payload must be bytes");
            case "Str": return NexusValue.Str(Convert.ToString(payload) ?? "");
            case "Array":
                if (payload is List<object?> larr)
                {
                    var outArr = new List<NexusValue>(larr.Count);
                    foreach (var e in larr) outArr.Add(FromWire(e));
                    return NexusValue.Array(outArr);
                }
                throw new InvalidOperationException("decode: Array payload must be list");
            case "Map":
                if (payload is List<object?> lmap)
                {
                    var pairs = new List<(NexusValue, NexusValue)>(lmap.Count);
                    foreach (var e in lmap)
                    {
                        if (e is List<object?> kv && kv.Count == 2)
                            pairs.Add((FromWire(kv[0]), FromWire(kv[1])));
                        else
                            throw new InvalidOperationException("decode: Map entry must be [key, value] pair");
                    }
                    return NexusValue.Map(pairs);
                }
                throw new InvalidOperationException("decode: Map payload must be list");
        }
        throw new InvalidOperationException($"decode: unknown NexusValue tag '{tag}'");
    }

    /// <summary>The decoded wire-level response.</summary>
    public class RpcResponse
    {
        public uint Id { get; set; }
        public bool Ok { get; set; }
        public NexusValue Value { get; set; }
        public string Err { get; set; } = "";

        public NexusValue Unwrap() =>
            Ok ? Value : throw new InvalidOperationException($"server: {Err}");
    }

    /// <summary>The wire-level request frame.</summary>
    public class RpcRequest
    {
        public uint Id { get; set; }
        public string Command { get; set; } = "";
        public List<NexusValue> Args { get; set; } = new();
    }

    /// <summary>
    /// Encode a request into a length-prefixed MessagePack frame:
    /// <c>u32_le(body_len) ++ msgpack(body)</c>.
    /// </summary>
    public static byte[] EncodeRequestFrame(RpcRequest req)
    {
        var args = new List<object?>(req.Args.Count);
        foreach (var a in req.Args) args.Add(ToWire(a));
        var body = MessagePackSerializer.Typeless.Serialize(new Dictionary<object, object?>
        {
            ["id"] = req.Id,
            ["command"] = req.Command,
            ["args"] = args,
        });

        var frame = new byte[4 + body.Length];
        frame[0] = (byte)(body.Length & 0xFF);
        frame[1] = (byte)((body.Length >> 8) & 0xFF);
        frame[2] = (byte)((body.Length >> 16) & 0xFF);
        frame[3] = (byte)((body.Length >> 24) & 0xFF);
        Buffer.BlockCopy(body, 0, frame, 4, body.Length);
        return frame;
    }

    /// <summary>
    /// Decode a response body (MessagePack bytes **after** the length
    /// prefix).
    /// </summary>
    public static RpcResponse DecodeResponseBody(byte[] body)
    {
        var raw = MessagePackSerializer.Typeless.Deserialize(body);
        if (raw is not IDictionary<object, object?> map)
            throw new InvalidOperationException($"decode: response must be a map, got {raw?.GetType().Name}");

        if (!map.TryGetValue("id", out var idRaw))
            throw new InvalidOperationException("decode: response missing 'id'");
        var id = Convert.ToUInt32(idRaw);

        if (!map.TryGetValue("result", out var resultRaw))
            throw new InvalidOperationException("decode: response missing 'result'");
        if (resultRaw is not IDictionary<object, object?> resultMap || resultMap.Count != 1)
            throw new InvalidOperationException("decode: Result must be a single-key tagged map");

        foreach (var kv in resultMap)
        {
            var tag = kv.Key as string ?? kv.Key.ToString();
            switch (tag)
            {
                case "Ok":
                    return new RpcResponse { Id = id, Ok = true, Value = FromWire(kv.Value) };
                case "Err":
                    return new RpcResponse { Id = id, Ok = false, Err = Convert.ToString(kv.Value) ?? "" };
                default:
                    throw new InvalidOperationException(
                        $"decode: Result must be 'Ok' or 'Err', got '{tag}'");
            }
        }
        throw new InvalidOperationException("decode: empty result map");
    }
}

File: sdks/typescript/src/transports/codec.ts
/**
 * Wire codec for the Nexus RPC protocol.
 *
 * The server-side types live in `nexus-protocol::rpc::types`. rmp-serde
 * encodes Rust enums using the externally-tagged representation by
 * default, which means:
 *
 * - Unit variants (`NexusValue::Null`) → the string `"Null"`.
 * - Data-bearing variants (`NexusValue::Str("hi")`) → a single-key map
 *   `{"Str": "hi"}`.
 * - `Result<T, E>` follows the same rule: `{"Ok": v}` / `{"Err": s}`.
 *
 * `toWireValue` / `fromWireValue` translate between the tagged
 * TypeScript type and that on-wire shape. `encodeRequest` and
 * `decodeResponse` handle the `u32 LE length prefix + MessagePack`
 * frame format documented in `docs/specs/rpc-wire-format.md`.
 */

import { pack, unpack } from 'msgpackr';
import { NexusValue, nx, TransportResponse } from './types';

/** Request frame serialised onto the wire. */
export interface RpcRequest {
  id: number;
  command: string;
  args: NexusValue[];
}

/** Decoded response — either a value or an error string. */
export type RpcResponseResult =
  | { ok: true; value: NexusValue }
  | { ok: false; message: string };

export interface RpcResponse {
  id: number;
  result: RpcResponseResult;
}

/**
 * Encode a `NexusValue` into its on-wire (pre-MessagePack) JS shape.
 *
 * Returns the literal string `'Null'` for unit variants and a single-key
 * object `{ VariantName: payload }` for every data-bearing variant.
 * Bytes are encoded as `Uint8Array` — msgpackr preserves them verbatim.
 */
export function toWireValue(v: NexusValue): unknown {
  switch (v.kind) {
    case 'Null':
      return 'Null';
    case 'Bool':
      return { Bool: v.value };
    case 'Int':
      // msgpackr encodes both number and bigint as integer types —
      // bigint is required when the magnitude exceeds 2^53-1.
      return { Int: typeof v.value === 'bigint' ? v.value : BigInt(v.value) };
    case 'Float':
      return { Float: v.value };
    case 'Bytes':
      return { Bytes: v.value };
    case 'Str':
      return { Str: v.value };
    case 'Array':
      return { Array: v.value.map(toWireValue) };
    case 'Map':
      // rmp-serde encodes `Vec<(K, V)>` as an array of 2-tuples — NOT a
      // msgpack map, since keys can be non-string NexusValues.
      return { Map: v.value.map(([k, val]) => [toWireValue(k), toWireValue(val)]) };
  }
}

/** Decode the wire-level JS shape back into a tagged `NexusValue`. */
export function fromWireValue(raw: unknown): NexusValue {
  if (raw === 'Null' || raw === null || raw === undefined) {
    return nx.Null();
  }
  if (typeof raw !== 'object') {
    // msgpackr can surface primitive variants as primitives when the
    // server sends a pre-tagged value through a different codepath.
    // Fall back to the most specific tag.
    if (typeof raw === 'boolean') return nx.Bool(raw);
    if (typeof raw === 'string') return nx.Str(raw);
    if (typeof raw === 'bigint') return nx.Int(raw);
    if (typeof raw === 'number') {
      return Number.isInteger(raw) ? nx.Int(raw) : nx.Float(raw);
    }
    return nx.Null();
  }

  // Bytes surface as Uint8Array (or Buffer on Node). Check before the
  // plain-object branch.
  if (raw instanceof Uint8Array) {
    return nx.Bytes(raw);
  }

  if (Array.isArray(raw)) {
    // Untagged array — interpret as `Array` variant.
    return nx.Array(raw.map(fromWireValue));
  }

  const entries = Object.entries(raw as Record<string, unknown>);
  if (entries.length !== 1) {
    throw new Error(`decode: expected single-key tagged NexusValue, got ${entries.length} keys`);
  }
  const [tag, payload] = entries[0];
  switch (tag) {
    case 'Null':
      return nx.Null();
    case 'Bool':
      return nx.Bool(Boolean(payload));
    case 'Int': {
      if (typeof payload === 'bigint') return nx.Int(payload);
      if (typeof payload === 'number') return nx.Int(payload);
      throw new Error(`decode: Int payload must be number or bigint`);
    }
    case 'Float': {
      if (typeof payload === 'number') return nx.Float(payload);
      if (typeof payload === 'bigint') return nx.Float(Number(payload));
      throw new Error(`decode: Float payload must be numeric`);
    }
    case 'Bytes': {
      if (payload instanceof Uint8Array) return nx.Bytes(payload);
      if (Array.isArray(payload)) return nx.Bytes(Uint8Array.from(payload as number[]));
      throw new Error(`decode: Bytes payload must be bytes`);
    }
    case 'Str':
      return nx.Str(String(payload));
    case 'Array': {
      if (!Array.isArray(payload)) throw new Error(`decode: Array payload must be array`);
      return nx.Array(payload.map(fromWireValue));
    }
    case 'Map': {
      if (!Array.isArray(payload)) throw new Error(`decode: Map payload must be array`);
      return nx.Map(
        payload.map((pair) => {
          if (!Array.isArray(pair) || pair.length !== 2) {
            throw new Error(`decode: Map entry must be [key, value] pair`);
          }
          return [fromWireValue(pair[0]), fromWireValue(pair[1])] as [NexusValue, NexusValue];
        })
      );
    }
    default:
      throw new Error(`decode: unknown NexusValue tag '${tag}'`);
  }
}

/**
 * Encode a request into a length-prefixed MessagePack frame.
 *
 * Wire layout: `u32_le(body_len) ++ msgpack(body)`.
 */
export function encodeRequestFrame(req: RpcRequest): Uint8Array {
  const body = pack({
    id: req.id,
    command: req.command,
    args: req.args.map(toWireValue),
  });
  const frame = new Uint8Array(4 + body.length);
  const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);
  view.setUint32(0, body.length, true);
  frame.set(body, 4);
  return frame;
}

/**
 * Decode a response body (the MessagePack-encoded bytes **after** the
 * length prefix). The caller is responsible for reading exactly
 * `length` bytes off the wire before handing them here.
 */
export function decodeResponseBody(body: Uint8Array): RpcResponse {
  const raw = unpack(body) as { id: number | bigint; result: unknown };
  const id = typeof raw.id === 'bigint' ? Number(raw.id) : Number(raw.id);
  const result = decodeResultEnvelope(raw.result);
  return { id, result };
}

function decodeResultEnvelope(raw: unknown): RpcResponseResult {
  if (raw === null || typeof raw !== 'object') {
    throw new Error(`decode: Result must be a tagged map, got ${typeof raw}`);
  }
  const entries = Object.entries(raw as Record<string, unknown>);
  if (entries.length !== 1) {
    throw new Error(`decode: Result must have exactly one tagged key`);
  }
  const [tag, payload] = entries[0];
  switch (tag) {
    case 'Ok':
      return { ok: true, value: fromWireValue(payload) };
    case 'Err':
      return { ok: false, message: String(payload) };
    default:
      throw new Error(`decode: Result must be 'Ok' or 'Err', got '${tag}'`);
  }
}

/** Build a `TransportResponse` from the server-side `Response` frame. */
export function rpcResponseToTransport(resp: RpcResponse): TransportResponse {
  if (!resp.result.ok) {
    throw new Error(`server: ${resp.result.message}`);
  }
  return { value: resp.result.value };
}
